### PR TITLE
Fix compilation for large tensor with MKL

### PR DESCRIPTION
**Pull request by anko-intel**
It fixes the issue #18954
The main changes:

- define MKL_INT to be consistent with mshadow index_t definition; This way LAPACKE  functions declaration in mkl/include/mkl_lapacke.h use the same type as MxNet is used in mshadow::index_t 
- define lapack_index_t to use int64_t as index type only for MKL
- define IndexT to use it in common code for CPU and GPU implementations. This way, when -DUSE_CUDA=1 -DUSE_INT64_TENSOR_SIZE=ON -DUSE_BLAS=mkl are set,  we have 2 different instantiation for GPU and CPU.
- convert the  pivot tensor for det and slodget operators from/to int64_t on GPU path to have common API with CPU
- eliminate needs to set MKL_USE_ILP64  separately, and force to set it consistently with USE_INT64_TENSOR_SIZE

### CODE REVIEW

**Comment by Zha0q1** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by Zha0q1** [comment]
@mxnet_bot run ci [centos_gpu]

**Comment by anko-intel** [comment]
@mxnet-bot run ci [centos-gpu, unix-gpu]

**Comment by Zha0q1** [comment]
> @mxnet-bot run ci [centos-gpu, unix-gpu]

Looks like it was just being flacky

**Comment by anko-intel** [comment]
@mxnet-bot run ci [centos-cpu]

**Comment by access2rohit** [comment]
re-triggerred sanity

**Comment by anko-intel** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by Zha0q1** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by Zha0q1** [comment]
@anko-intel could you check the windows build? Thanks!

**Comment by leezu** [comment]
@Zha0q1 @anko-intel the windows gpu failure is to an infamous bug in Cuda 10. It is usually mitigated by retrying compilation for 5 times, but your PR was unlucky or the code change has increased the probability of the Cuda compiler bug.

@josephevans is helping to update the CI to Cuda 11 to finally get rid of the bug. CC @sandeep-krishnamurthy 

@mxnet-bot run ci [windows-gpu]

**Comment by Zha0q1** [comment]
> @Zha0q1 @anko-intel the windows gpu failure is to an infamous bug in Cuda 10. It is usually mitigated by retrying compilation for 5 times, but your PR was unlucky or the code change has increased the probability of the Cuda compiler bug.
> 
> @josephevans is helping to update the CI to Cuda 11 to finally get rid of the bug. CC @sandeep-krishnamurthy
> 
> @mxnet-bot run ci [windows-gpu]

Ohh I see. Thanks for the clarification!

**Comment by Zha0q1** [comment]
> @Zha0q1 @anko-intel the windows gpu failure is to an infamous bug in Cuda 10. It is usually mitigated by retrying compilation for 5 times, but your PR was unlucky or the code change has increased the probability of the Cuda compiler bug.
It's failing again. We are talking about this issue right?
```
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2993: 'T': illegal type for non-type template parameter '__formal'
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): note: see reference to class template instantiation 'thrust::detail::allocator_traits_detail::has_system_type<T>' being compiled
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2065: 'U1': undeclared identifier
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2923: 'std::_Select<__formal>::_Apply': 'U1' is not a valid template type argument for parameter '<unnamed-symbol>'
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2144: syntax error: 'unknown-type' should be preceded by ')'
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2144: syntax error: 'unknown-type' should be preceded by ';'
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2238: unexpected token(s) preceding ';'
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2059: syntax error: ')'
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2988: unrecognizable template declaration/definition
[2020-11-06T06:14:35.175Z] C:/Windows/TEMP/tmpkffgnho0/thrust-1.9.8\thrust/detail/allocator/allocator_traits.h(54): error C2059: syntax error: '<end Parse>'
[2020-11-06T06:14:35.175Z] ../3rdparty/tvm/nnvm/include\nnvm/symbolic.h(73): warning C4251: 'nnvm::Symbol::outputs': class 'std::vector<nnvm::NodeEntry,std::allocator<nnvm::NodeEntr
```
that's really an interesting bug :p

**Comment by anko-intel** [comment]
@mxnet-bot  run ci [windows-gpu]

File: src/operator/tensor/la_op-inl.h [hunk 1]
```diff
@@ -931,15 +1020,20 @@ struct det_backward {
...
+                                     reinterpret_cast<const Tensor<xpu, 2, IndexInternalT>&>(pivot),
+                                     det, dA, DType(0), ctx);
     const_cast<Tensor<xpu, 3, DType>&>(dA) = broadcast_to(reshape(det * ddet, \
       Shape3(det.size(0), 1, 1)), mxnet::TShape(LU.shape_)) * \
       transpose(LU, Shape3(0, 2, 1));
-    Stream<xpu> *s = ctx.get_stream<xpu>();
     // stop grad for zero det temporarily
     Kernel<StopZeroDetGrad, xpu>::Launch(s, dA.shape_.Size(), dA.size(1) * dA.size(2), \
                                          dA.dptr_, det.dptr_, DType(0));
+    convert_to_int64_if_needed(s, pivot);
```
Comment (Zha0q1) [inline src/operator/tensor/la_op-inl.h:1036]: I think the only output is dA? In that case we might not need to convert the results back to int64

**Comment by leezu** [comment]
@Zha0q1 yes, that's the bug. NVCC is non-deterministically producing invalid output causing the MSVC to fail. You can find more details at https://github.com/thrust/thrust/issues/1090

**Comment by leezu** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by leezu** [comment]
@anko-intel can you try to rebase and force-push?

I'm concerned about merging this PR as for unknown reason it makes the nvcc bug more frequent. Maybe @josephevans can provide an ETA for the Windows Cuda 11 work and we may be able to wait for that?

**Comment by Zha0q1** [comment]
Regarding the type conversion on the gpu path for operator `det` and `slogdet`, maybe we can use a kernel like this: 
```c++
struct CopyArray {
  template<typename SType, typename DType>
  MSHADOW_XINLINE static void Map(size_t i, SType* src, DType* dest) {
    dest[i] = src[i];
  }
};

    ......
    // in det forward 
    if (std::is_same<xpu, gpu>::value && !std::is_same<IndexT, int>::value) {
      using IndexInternalT = typename LapackIndex<xpu>::IndexT;
      Tensor<xpu, 2, IndexInternalT> workspace =
          ctx.requested[0].get_space_typed<xpu, 2, IndexInternalT>(pivot.shape_, s);
      linalg_batch_getrf(LU, workspace, false, s);
      Kernel<CopyArray, xpu>::Launch(s, pivot.shape_.Size(), workspace.dptr_, pivot.dptr_);
    } else {
      linalg_batch_getrf(LU, pivot, false, s);
    }
```

**Comment by Zha0q1** [comment]
```c++
template<
  typename xpu,
  typename IndexT,
  std::enable_if_t<!std::is_same<IndexT, lapack_index_t>::value, int> = 0>
inline void convert_to_int_if_needed(
              Stream<xpu> *s,
              const Tensor<xpu, 2, IndexT>& tensor) {
}

// convertion to int is required only for GPU when IndexT is equal lapack_index_t (int64_t)
template<
  typename xpu,
  typename IndexT,
  std::enable_if_t<std::is_same<IndexT, lapack_index_t>::value, int> = 0>
inline void convert_to_int_if_needed(
              Stream<xpu> *s,
              const Tensor<xpu, 2, IndexT>& tensor) {
#ifdef __CUDACC__
  CHECK_LE(tensor.shape_[0], std::numeric_limits<int>::max())
    << "Tensor has size greater than supported.";
  CHECK_LE(tensor.shape_[1], std::numeric_limits<int>::max())
    << "Tensor has size greater than supported.";
  cudaStream_t stream = Stream<xpu>::GetStream(s);
  size_t elements =  tensor.shape_.Size();
  std::vector<IndexT> vec(elements, 0);
  IndexT* ptr = vec.data();
  int* ptr_int = reinterpret_cast<int*>(vec.data());

  CUDA_CALL(cudaMemcpyAsync(ptr, reinterpret_cast<IndexT*>(tensor.dptr_),
                            tensor.MSize() * sizeof(IndexT),
                            cudaMemcpyDeviceToHost, stream));
  for (IndexT i = 0; i < elements; ++i) {
    ptr_int[i] = static_cast<int>(ptr[i]);
  }
  CUDA_CALL(cudaMemcpyAsync(tensor.dptr_, ptr,
                            tensor.MSize() * sizeof(IndexT),
                            cudaMemcpyHostToDevice, stream));
#endif
}
```

This might be the cause of the windows build issue if I were to take a guess.

File: src/operator/tensor/la_op-inl.h (hunk 1 above)
Comment (anko-intel) [reply src/operator/tensor/la_op-inl.h:1036]: You are right. I had doubt if we can modify memory pointed by the input tensor. I will remove this line

**Comment by anko-intel** [comment]
Thank you @Zha0q1 for your comments. I measured performance with your kernel from commit: https://github.com/apache/incubator-mxnet/pull/19489/commits/31e07c4421aaa58a4812d42e9964af7986356b45 and it gives better results that my original https://github.com/apache/incubator-mxnet/pull/19067/commits/fd624bea7f2a9d38605a275e75c6112a71cdd0d2:
image
So I put it as solution for dot and slogdet on GPU path.

**Comment by anko-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by anko-intel** [comment]
@mxnet-bot run ci [windows-cpu]

**Comment by Zha0q1** [comment]
PR is ready! Would you review @leezu

**Review by leezu — APPROVED** [review]
Thank you!

**Pull request merged**